sentry/src/db/campaign.rs

```diff
@@ -54,6 +59,114 @@ pub async fn fetch_campaign(
+    }
+
+    // Use a varchar or text, since otherwise `int8` fails deserialization
+    accepts!(VARCHAR, TEXT);
+}
+
+pub async fn list_campaigns(
+    pool: &DbPool,
+    skip: u64,
+    limit: u32,
+    creator: &Option<Address>,
+    validator: &Option<ValidatorId>,
+    is_leader: &Option<bool>,
+    active_to_ge: &DateTime<Utc>,
+) -> Result<CampaignListResponse, PoolError> {
+    let client = pool.get().await?;
+
+    let validator = validator.as_ref().map(|validator_id| {
+        serde_json::Value::from_str(&format!(r#"[{{"id": "{}"}}]"#, validator_id))
+            .expect("Not a valid json")
+    });
+    let (where_clauses, params) =
+        campaign_list_query_params(creator, validator.as_ref(), is_leader, active_to_ge);
+    let total_count_params = (where_clauses.clone(), params.clone());
+
+    // To understand why we use Order by, see Postgres Documentation: https://www.postgresql.org/docs/8.1/queries-limit.html
+    let statement = format!("SELECT id, channel, creator, budget, validators, title, pricing_bounds, event_submission, exhausted, ad_units, targeting_rules, created, active_from, active_to FROM campaigns WHERE {} ORDER BY spec->>'created' DESC LIMIT {} OFFSET {}", where_clauses.join(" AND "), limit, skip);
+    let stmt = client.prepare(&statement).await?;
+
+    let rows = client.query(&stmt, params.as_slice()).await?;
+    let campaigns = rows.iter().map(Campaign::from).collect();
+
+    let total_count =
+        list_campaigns_total_count(pool, (&total_count_params.0, total_count_params.1)).await?;
+
+    // fast ceil for total_pages
+    let total_pages = if total_count == 0 {
+        1
+    } else {
+        1 + ((total_count - 1) / limit as u64)
+    };
+
+    let pagination = Pagination {
+        total_pages,
+        total: total_pages,
+        page: skip / limit as u64,
+    };
+
+    Ok(CampaignListResponse {
+        pagination,
+        campaigns,
+    })
+}
+
+fn campaign_list_query_params<'a>(
+    creator: &'a Option<Address>,
+    validator: Option<&'a serde_json::Value>,
+    is_leader: &Option<bool>,
+    active_to_ge: &'a DateTime<Utc>,
+) -> (Vec<String>, Vec<&'a (dyn ToSql + Sync)>) {
+    let mut where_clauses = vec!["active_to >= $1".to_string()];
+    let mut params: Vec<&(dyn ToSql + Sync)> = vec![active_to_ge];
+
+    if let Some(creator) = creator {
+        where_clauses.push(format!("creator = ${}", params.len() + 1));
+        params.push(creator);
+    }
+
+    // TODO: if no validator is provided as query param, validator will be the one from Auth, is that ok?
+    if let Some(validator) = validator {
+        where_clauses.push(format!("validators @> ${}", params.len() + 1));
+        params.push(validator);
+    }
+
+    if let (Some(validator), Some(true)) = (validator, is_leader) {
+        where_clauses.push(format!("channel->'leader' = ${}", params.len() + 1)); // TODO: change after channel is in a different table
+        params.push(validator); // TODO: could be redundant?
```

Comment:
Yes, it porbablly is. I would suggest to have this check in 1 place (i.e. above) because for the `validator` & `is_leader` you will require the same parameter - ValidatorId. Only 1 `params.push()` is required, but also the same `param_num = params.len() + 1` for the validator & is_leader checks